Include challenge password attribute if required by EST server

mobe1 opened the pull request:
The changes introduced come after the issue/feature requrest https://github.com/globalsign/est/issues/30 has been opened.
They allow us to enroll a CSR that includes the TLS-unique value as recommended by the RFC 7030

- Because each http client instantiation results in a new TLS-unique, one way of including it would be to make EST requests from the same http client.
- Because the standard crypto/x509 Go package does not handle the challenge password attribute (OID) the way an EST/CA server expects it, the CSR creation had to be wrapped.

mobe1 committed:
Make all http requests from same http client

Code refactoring : change Enroll(), Reenroll() and ServerKeyGen() csr argument type so that we don't depend on x509 package anymore, which today still ignores the challenge password attribute

Inlcude tls unique if required by CA

Add test cases

Add sample

Update readme and documentation

toddgaunt-gs commented:
Thanks for opening this PR, I'll forward this to my team for review.

toddgaunt-gs commented:
Need to resolve conflicts after upgrading to Go 1.22.1...

mobe1 committed:
Merge remote-tracking branch 'upstream/master' into features/include_tlsunique

mobe1 commented:
Thought the comment wasn't for me until I noticed the repo did get upgraded with a different version of go...  
The conflicts should be resolved.

toddgaunt-gs commented:
Ah it wasn't actually! Just a reminder for when I had time to get around to this again, but thank you for addressing the changes needed!

DDvO commented:
Looks like work on this PR has stalled - I wonder when is it going to be resumed?

DDvO commented:
BTW, copying to `csr.go` large chunks of code from the `x509 package` does not look right -
there should be a lean way of adding and reading the `challengePassword` attribute of a PKCS#10 CSR.

toddgaunt-gs commented:
Hi DDvO, unfortunately my team hasn't been able to spare the cycles to work on this repository for a while, so while I hesitate to recommend it you may want to fork and patch a clone of this repository if you need this functionality right away.

toddgaunt-gs commented:
> BTW, copying to `csr.go` large chunks of code from the `x509 package` does not look right - there should be a lean way of adding and reading the `challengePassword` attribute of a PKCS#10 CSR.

I agree with this, I'd prefer to see this functionality without including this dependency if possible.

toddgaunt-gs commented:
Sorry for such a long delay for reviewing this PR, I didn't prioritize it has been far to long so I apologize for that. Thank you for your contribution, but I don't think I can hit the merge button on this PR as it is now into the master, but I am happy to support it as a separate branch to support this use-case for now.

Some critiques that are blocking this:
- I'd prefer the smaller style changes, such as using `io` over `ioutil` and changing the code to call `EqualFold` be done in a separate PR as these changes are unrelated to the challenge password.
- As @DDvO mentions above, avoiding copying csr.go would be preferred
- The samples and resources directory added I'll have to give some more thought to. While I like seeing a way to try out the changes, I think I'd prefer this just as an automated test of the feature instead. I think adding the case to est_test.go would be the way to go.

mobe1 commented:
@toddgaunt-gs 
Nah acutally, you're right. I shouldn't have mixed code refactoring and the feature itself as it doesn't help telling what's necessary to make the feature available...

On a personal note: I thought I had to change the method signature but after a quick review, I start to think it can be done while keeping your methods signatures intact.

That being said,
- I'll revert those unrelated changes
- then I'll bring the method signature back
- finally, i'll see if there is a better way of adding that `tls-unique` without copying the `csr.go` (spoiler alert, i'm not very confident on this one)

DDvO commented:
Hi guys,

thank you for swiftly responding on this.

> > BTW, copying to csr.go large chunks of code from the x509 package does not look right -
> > there should be a lean way of adding and reading the challengePassword attribute of a PKCS#10 CSR.
> 
> I too agree, I actually borrowed that code snippet from another repository, credit goes to [micromdm](https://github.com/globalsign/est/pull/38/files#diff-e7af9b167346dc2be28de5632d020db09219b84b9324531b935331b838745c3cR51).

Looks like there is a glitch in the link you included and instead you meant
https://github.com/micromdm/scep/blob/main/cryptoutil/x509util/x509util.go

> Meanwhile, unfortunately Go has still not taken action to address it, there is an [open issue](https://github.com/golang/go/issues/15995) for it already.

I've meanwhile had a closer look.
Oh my goodness. The `crypto/x509` package screwed up the implementation of the [PKCS#10 CSR `Attribute` type]([url](https://datatracker.ietf.org/doc/html/rfc2986#section-4)),
which is hard to work around also due to its very restrictive API.
I just [commented there](https://github.com/golang/go/issues/15995#issuecomment-2792691214) accordingly.

mobe1 committed:
refactor: preserve method signature (#2)

* add challenge password to CSR as described in rfc 7030 section 3.5
* update cmd client test : if challenge password is to be included, CSR must be re-created.

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: Rob Casey <rcasey@gmail.com>
Co-authored-by: Todd Gaunt <todd.gaunt@globalsign.com>
Co-authored-by: Rob Casey <61131@users.noreply.github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

mobe1 committed:
Merge remote-tracking branch 'upstream/master' into features/include_tlsunique

mobe1 committed:
Merge remote-tracking branch 'refs/remotes/origin/features/include_tlsunique' into features/include_tlsunique

mobe1 committed:
Merge remote-tracking branch 'upstream/master' into features/include_tlsunique

mobe1 committed:
revert unrelated change to the challenge password feature

mobe1 commented:
Hey guys,

It appears Go hasn't moved much since, aside from labeling it `NeedsInvestigation`.
@DDvO, I checked out the [CSR tool](https://github.com/golang/go/issues/15995#issuecomment-2800868263) project your referenced as well
- it does add the challenge password to the CSR
- however, that's the only thing it does (most of the CSR fields are still missing)

Also noticed the project hasn't moved since its creation date and the fact he did not respond to your remark aren't very good signs.

At this point, I don't know if there are better alternatives
- I took the shortcut of using existing code from [micromdm](https://github.com/micromdm/scep/tree/main) `scep` package that already "solved" the problem, which implied a huge code duplication from `x509` package though
- Wait for Go to act on it and fix the CSR attribute implementation

toddgaunt-gs commented on csr.go:
```diff
@@ -0,0 +1,374 @@
+// Copied from https://github.com/micromdm/scep/blob/main/cryptoutil/x509util/x509util.go#L58
+func CreateCertificateRequest(rand io.Reader, template *CertificateRequest, priv any) (csr []byte, err error) {
+	if template == nil {
+		return nil, errors.New("make sure to rovide a certificate request template")
```
The error string here has a typo: "make sure to _rovide_..." should be "provide"

toddgaunt-gs reviewed (changes requested):
Unfortunately after further review I still don't think I'd like to merge this PR as is. Thank you for addressing prior feedback, but the following changes would need to be made before I merge this PR into the master branch.:

- Review comments addressed
- The `sample` program removed or rewritten into a directory containing a config file for the EST client and either a shell script or instructions for how to operate it for the use case it is demonstrating
- The copied challenge password code found in parts of csr.go isolated into its own package with a clear README file indicating the source without any new or custom code included alongside it.

I understand that these changes I'm asking for may not be a priority for you as you have a functional fork of this repository working for you, and updates to this one are few and far between. Hopefully this review can at least benefit your own fork if you do decide you want to pursue these changes. Let me know if so and we can discuss in more detail each point to avoid misunderstanding any of the requested changes.

mobe1 reviewed (commented):
Yes, i did miss a critical detail on `PoP` in the context of `mTLS`. I'll get back to it as soon as i can.
More details on the corresponding comment.

As for the `/sample` directory, I'll just drop it cause there is no straightforward way to show it added the `challengePassword`.
I used to verify it via the EST server. It was easy and quick for me to just `go run` it.

mobe1 committed:
Remove the sample from the repository

toddgaunt-gs commented on client.go:
```diff
@@ -447,6 +485,21 @@ func (c *Client) newRequest(
+		CertificateRequest: *standardLibCsr,
+		ChallengePassword:  challengePassword,
+	}
+	crBs, err := CreateCertificateRequest(rand.Reader, &cr, c.PrivateKey)
```
I agree, I think a new field that can be set through a command-line flag would work best. I think `SigningKey` and `signing-key` would work well respectively. Importantly if the field is nil/the flag is not provided, we should not re-sign the CSR which would result in not using the tls-unique value for the challenge-password, but I think that is preferred behavior.

mobe1 committed:
remove .md warnings related to the syntax

mobe1 committed:
introduce signing key field for signing CSR when challenge password is included

mobe1 committed:
take into account signing key in est client CLI

mobe1 committed:
use smallstep scep package to add challenge password instead of copy pasting

mobe1 committed:
readme : enforce pop as per RFC 7030

mobe1 committed:
go mod update

mobe1 committed:
Merge branch 'master' into features/include_tlsunique

mobe1 commented:
Hi @toddgaunt-gs, how you doing :)

Sorry for the delay, I had some time off and realized I haven't updated my PR since.

I closed the obvious comments : 
- I directly import the package that includes the tls-unique/challenge password (instead of copy/pasting with a comment).
- The sample directoy was removed as it doesn't provide any value to the repo itself.

Finally, I fixed the defect I was about to introduce, I left the comment open for reminding the context and the agreed solution.
I also updated the EST client CLI along the readme.

Let me know if something else needs to be done.

mobe1 commented on README.md:
```diff
@@ -9,18 +9,18 @@ enrollment protocol as defined by [RFC7030](https://tools.ietf.org/html/rfc7030)
- * The optional additional path segment
- * Optional HTTP-based client authentication on top of certificate-based
+* The optional additional path segment
+* Optional HTTP-based client authentication on top of certificate-based
```
Juts a reminder for me -> undo this change related to single space removal in each line.

mobe1 commented on est_test.go:
```diff
@@ -381,7 +380,9 @@ func TestReenroll(t *testing.T) {
-			// Create test EST server and client.
+			// Create test EST client.
+			client := newTestClient()
+			client.PrivateKey = tc.key
```
Reminder for me -> can be reverted without any impact on unit tests since we make the difference between `private` and `signing` keys now

mobe1 commented on est_test.go:
```diff
@@ -601,6 +601,8 @@ func TestTPMEnroll(t *testing.T) {
 			ek := mustGenerateRSAPrivateKey(t)
 			csr := mustCreateCertificateRequest(t, ek, "Test TPM Device", nil)
+
+			client.PrivateKey = ek
```
Reminder for me -> can be reverted without any impact on unit tests since we make the difference between `private` and `signing` keys now

mobe1 reviewed (commented):
Noticed a few changes that are not longer required considering the latest solution that has been suggested/accepted.

I'll deal with them right away.

mobe1 committed:
revert remaining unrelated changes

mobe1 committed:
godoc update

mobe1 commented:
@toddgaunt-gs hi,
I think i have reverted all of the unrelated changes.

I saw there hasn't been a tag yet in one of your packages.
=> I used the commit hash instead in the meantime.

And, i'm not sure if it's relevant, but just in case, I made sure I didn't break anything in the CLI.
I used different options and checked the csr `signing-key` did not interfere with the mTLS private `key`.

<img width="1742" height="121" alt="image" src="https://github.com/user-attachments/assets/28a2094e-d0f3-4ca0-a350-9483ed397515" />

I tested with a different CA to verify the final CSR the `estclient` CLI sent did include the challenge password.

<img width="506" height="299" alt="image" src="https://github.com/user-attachments/assets/5243d9ee-20f1-4428-a553-0215177256d9" />

Tried with your testing server .exe as well, and it works just fine.
It's just I don't have a direct way of checking whether it includes any challenge password or not.

toddgaunt-gs commented:
Thanks for pinging me, sorry I'd been busy and haven't gotten back around to this. I'll take a look today.

toddgaunt-gs commented on client.go:
```diff
@@ -447,6 +492,22 @@ func (c *Client) newRequest(
+		return csr, nil
+	}
+
+	stdCsr, _ := x509.ParseCertificateRequest(csr)
```
This error shouldn't be ignored since the CSR can originate from something the user passes in. If this fails, it may result in a panic on line 504.

mobe1 replied:
Oh, that is so true !
I didn't notice `smallstep` was using the value instead of the pointer.

Good catch, i'm on it.

toddgaunt-gs commented on go.sum:
Would you mind running `go tidy` on this? I see quite a few changes when I do locally, mostly removing things we don't need anymore.

mobe1 replied:
Huh, I was sure i did run `go mod tidy`.
Let me try again.

mobe1 replied:
You're right, i did miss a `tidy`...

toddgaunt-gs commented:
LGTM outside of a few small things. Feel free to address them if you have the time, though they aren't critical. I'll request another coworker to review in the meanwhile.

mobe1 reviewed (commented):
Welcome back.

mobe1 committed:
return error early when parsing certificate to prevent go from panic

mobe1 commented:
Hi @mara-soldan welcome :)

I don't know when you'll get the time to review the PR.
But here's a little briefing of what has been done if it helps
- I took advantage of the existing `makeHTTPClient` to get the `tls-unique` session/value (aka, `challenge password` in EST specs)
- I imported the `smallstep` SCEP package which handles the `challengePassword` attribute in a CSR
  - it basically wraps the standard `x509.CertificateRequest` which does not handle it correctly
  - once the [issue](https://github.com/golang/go/issues/15995) is resolved in the framework Go, we can then remove the dependency to `smallstep` custom function
- And I updated Globalsign EST client CLI as well

And of course, if you see something else, let me know and I'll address it as soon as I can.

toddgaunt-gs commented:
@mobe1 merged since I didn't want to hold this up needlessly. Thanks for the contribution, and for being patient and pleasant to work with during the review/feedback cycle!

mobe1 commented:
@toddgaunt-gs I can say the same.
Appreciate !
